core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -720,41 +718,36 @@ public void valueAdded(Object key) {
 
       public void valueIterated(Object key) {
          pendingSegments.compute(key, iteratorMapping);
-         if (ref.get() != null) {
-            completionListener.accept(ref.get());
+         Supplier<PrimitiveIterator.OfInt> segments = ref.get();
+         if (segments != null) {
+            ref.set(null);
+            completionListener.accept(segments);
          }
       }
 
-      private IntStream toStream(PrimitiveIterator.OfInt iter) {
-         return StreamSupport.intStream(Spliterators.spliteratorUnknownSize(iter,
-               Spliterator.DISTINCT | Spliterator.NONNULL), false);
-      }
-
       public void segmentsEncountered(Supplier<PrimitiveIterator.OfInt> segments) {
          // This code assumes that valueAdded and segmentsEncountered are not invoked concurrently and that all values
          // added for a response before the segments are completed.
          // The valueIterated method can be invoked at any point however.
          // See ClusterStreamManagerImpl$ClusterStreamSubscription.sendRequest where the added and segments are called into
-         AtomicBoolean shouldNotify = new AtomicBoolean(true);
+         ByRef<Supplier<PrimitiveIterator.OfInt>> segmentsToNotify = new ByRef<>(segments);
          Object key = currentKey.get();
          if (key != null) {
             pendingSegments.compute(key, (k, v) -> {
-               // This check is if iterator caught up to us
+               // The iterator has consumed all the keys, so there is no reason to wait and just notify of segment
+               // completion immediately
                if (currentKey.get() == null) {
                   return null;
                }
-               shouldNotify.set(false);
-               if (v == null) {
-                  return segments;
-               } else {
-                  // This is ugly, but we shouldn't hit this often (need a response with no entries and completed
-                  // segments and iterator that hasn't caught up)
-                  return () -> Stream.of(segments, v).flatMapToInt(s -> toStream(s.get())).iterator();
-               }
+               // We save previous segments if there were any. This should normally be null since if we iterate
+               // over an entry then it will null this. So this is mostly for segments that have no entries
```

Comment:
If all the segments that will ever be set in `segmentsToNotify` are empty, then I'd make that clearer. Something like "valueIterated notifies the completion of non-empty segments, but we need to notify the completion of empty segments here".

Reply:
Sure